lib/soak/src/bin/observer.rs

```diff
@@ -97,123 +131,233 @@ enum Status {
-        for query in config.queries {
-            let url = format!("{}/api/v1/query?query={}", config.prometheus, query.query);
-            let url = url.parse::<Url>().unwrap();
-            queries.push((query, url));
-        }
+impl fmt::Debug for TargetWorker {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("TargetWorker")
+            .field("run_id", &self.run_id)
+            .field("target_id", &self.target_id)
+            .field("url", &self.url.as_str())
+            .finish()
+    }
+}
 
+impl TargetWorker {
+    fn new(
+        snd: Sender<String>,
+        experiment_name: String,
+        target_id: String,
+        url: Url,
+        run_id: Arc<Uuid>,
+        variant: soak::Variant,
+    ) -> Self {
         Self {
-            vector_id: config.vector_id,
-            experiment_name: config.experiment_name,
-            variant: config.variant,
-            queries,
-            capture_path: config.capture_path,
+            snd,
+            experiment_name,
+            target_id,
+            url,
+            run_id,
+            variant,
         }
     }
 
+    #[instrument]
     async fn run(self) -> Result<(), Error> {
         let client: reqwest::Client = reqwest::Client::new();
 
-        let file = fs::File::create(self.capture_path)?;
-        let mut wtr = csv::Writer::from_writer(file);
         for fetch_index in 0..u64::max_value() {
-            for (query, url) in &self.queries {
-                let request = client.get(url.clone()).build()?;
-                let body = client
-                    .execute(request)
-                    .await?
-                    .json::<QueryResponse>()
-                    .await?;
-                debug!("body: {:?}", body.data);
-
-                if !body.data.result.is_empty() {
-                    let time = body.data.result[0].value.time;
-                    let value = body.data.result[0].value.value.parse::<f64>()?;
+            if TERMINATE.load(Ordering::Relaxed) {
+                info!("Received terminate signal");
+                break;
+            }
+            sleep(Duration::from_millis(307)).await;
+
+            let now_ms: u128 = SystemTime::now().duration_since(UNIX_EPOCH)?.as_millis();
+            let request = client.get(self.url.clone()).build()?;
+            let response = match client.execute(request).await {
+                Ok(resp) => resp,
+                Err(e) => {
+                    debug!(
+                        "Did not receive a response from {} with error: {}",
+                        self.target_id, e
+                    );
+                    continue;
+                }
+            };
+            let body = response.text().await?;
+            let metric_groups = prometheus_parser::parse_text(&body)?;
+            if metric_groups.is_empty() {
+                debug!("failed to request body: {:?}", body);
```

Comment:
Seems to be missing a word in the message here

Reply:
Ah yeah, maybe a little terse. I'd be happy to change the wording but it's cromulent English, I believe, aside from the no punctuation or capitalization.

Reply:
🤔 oh, is "request" the verb here? I read it and saw no verb, but can see it as "failed when requesting the body".

Reply:
Yessir.